src/zcl_utesthelp_domconstant_test.clas.abap

```diff
@@ -0,0 +1,153 @@
+        VALUE(rv_result)  TYPE string.
+ENDCLASS.
+
+
+CLASS zcl_utesthelp_domconstant_test IMPLEMENTATION.
+  METHOD setup.
+    mo_helper = NEW #( ).
+    mv_class_name = get_class_name( ).
+
+    mt_constants = mo_helper->determine_constants( mv_class_name ).
+
+    mv_domain_name = mo_helper->determine_domain_name( iv_class_name = mv_class_name
+                                                       is_constant   = mt_constants[ 1 ] ).
+
+    mt_domain_values = mo_helper->determine_domain_values( mv_domain_name ).
+  ENDMETHOD.
+
+  METHOD all_constants_found_in_domain.
+    LOOP AT mt_constants ASSIGNING FIELD-SYMBOL(<ls_constant>).
+      DATA(lv_value) = get_attribute_value( <ls_constant>-name ).
+
+      IF NOT line_exists( mt_domain_values[ domvalue_l = lv_value ] ).
+        cl_abap_unit_assert=>fail( |Value { <ls_constant>-name } not found in domain fix values| ).
+      ENDIF.
+    ENDLOOP.
+  ENDMETHOD.
+
+  METHOD class_is_abstract.
+    DATA(lo_class_description) = CAST cl_abap_classdescr( cl_abap_typedescr=>describe_by_name( mv_class_name ) ).
+
+    cl_abap_unit_assert=>assert_equals( exp = cl_abap_classdescr=>classkind_abstract
+                                        act = lo_class_description->class_kind ).
+  ENDMETHOD.
+
+  METHOD all_domain_values_as_constants.
+    LOOP AT mt_domain_values ASSIGNING FIELD-SYMBOL(<ls_domain_value>).
+      DATA(lv_value_found) = abap_false.
+
+      LOOP AT mt_constants ASSIGNING FIELD-SYMBOL(<ls_constant>).
+        DATA(lv_value) = get_attribute_value( <ls_constant>-name ).
+
+        IF <ls_domain_value>-domvalue_l = lv_value.
+          lv_value_found = abap_true.
+          EXIT.
+        ENDIF.
+      ENDLOOP.
+
+      cl_abap_unit_assert=>assert_true(
+          act = lv_value_found
+          msg = |Domain value { <ls_domain_value>-domvalue_l } not found as constant| ).
+    ENDLOOP.
+  ENDMETHOD.
+
+  METHOD constant_values_do_not_repeat.
+    DATA lt_values TYPE TABLE OF string.
+
+    LOOP AT mt_constants ASSIGNING FIELD-SYMBOL(<ls_constant>).
+      DATA(lv_value) = get_attribute_value( <ls_constant>-name ).
+
+      cl_abap_unit_assert=>assert_table_not_contains( line  = lv_value
+                                                      table = lt_values
+                                                      msg   = |Value { lv_value } was used in multiple constants| ).
+
+      INSERT lv_value INTO TABLE lt_values.
+    ENDLOOP.
+  ENDMETHOD.
+
+  METHOD all_constants_have_same_type.
+    LOOP AT mt_constants ASSIGNING FIELD-SYMBOL(<ls_constant>).
+      DATA(lv_domain_value) = mo_helper->determine_domain_name( iv_class_name = mv_class_name
+                                                                is_constant   = <ls_constant> ).
+
+      cl_abap_unit_assert=>assert_equals( exp = mv_domain_name
+                                          act = lv_domain_value
+                                          msg = |Constant { <ls_constant>-name } has an unexpected type| ).
```

Comment:
```suggestion
                                          msg = |Constant { <ls_constant>-name } must be typed <lv_domain_value>| ).
```
I do not believe it is _needed_, but I think it would improve the solution as leaving out information (or hiding somewhere) is not the "better" approach, I think.

Reply:
I get what you mean. The main message that the constant must be of a certain type isn't quite clear with the current implementation.